src/lib/tableProvider.ts

```diff
@@ -37,33 +43,41 @@ export function parquetDataFrame(from: AsyncBufferFrom, metadata: FileMetaData):
   const sortCache = new Map<string, Promise<number[]>>()
   const columnRanksCache = new Map<string, Promise<number[]>>()
   const data = new Array<ResolvableRow | undefined>(Number(metadata.num_rows))
-  const groups = new Array(metadata.row_groups.length).fill(false)
+
+  // virtual row groups are up to 1000 rows within row group boundaries
+  const groups: VirtualRowGroup[] = []
   let groupStart = 0
-  const groupEnds = metadata.row_groups.map(group => groupStart += Number(group.num_rows))
+  for (const rg of metadata.row_groups) {
+    // make virtual row groups of size 1000
+    for (let j = 0; j < rg.num_rows; j += 1000) {
+      const groupSize = Math.min(1000, Number(rg.num_rows) - j)
```

Comment:
why are you using Number here? in case it's a BigInt? Do we need to do the same in the previous line, or is not an issue?

Reply:
RowGroup.num_rows is always a bigint. The problem is you can't mix number and bigint in mathematical operations like `-` but you CAN mix them in comparison operations like `<`.

:heart: js